crates/red_knot_python_semantic/src/symbol.rs

```diff
@@ -44,15 +44,25 @@ impl<'db> Symbol<'db> {
         }
     }
 
-    pub(crate) fn unwrap_or(&self, other: Type<'db>) -> Type<'db> {
+    pub(crate) fn unwrap_or_unknown(&self) -> Type<'db> {
         match self {
             Symbol::Type(ty, _) => *ty,
-            Symbol::Unbound => other,
+            Symbol::Unbound => Type::Unknown,
         }
     }
 
-    pub(crate) fn unwrap_or_unknown(&self) -> Type<'db> {
-        self.unwrap_or(Type::Unknown)
+    /// Get the return type of a call to a symbol if it is definitely bound and callable.
+    ///
+    /// Returns `None` if the symbol is (possibly) unbound or if the type is not callable.
+    pub(crate) fn get_return_type_if_callable(
+        self,
+        db: &'db dyn Db,
+        arg_types: &[Type<'db>],
+    ) -> Option<Type<'db>> {
+        match self {
+            Symbol::Type(ty, Boundness::Bound) => ty.call(db, arg_types).return_ty(db),
+            Symbol::Type(_, Boundness::MayBeUnbound) | Symbol::Unbound => None,
+        }
```

Comment:
Discussion: Arguably, this method is not much better in terms of making its semantics clear. But `get_return_type_if_definitely_bound_and_callable` didn't sound very catchy.

And it is extremely useful for simplifying code at the call sites (three, so far). I tried various other ways of rewriting the part in `infer_binary_expression_type`, and nothing came close to what we have with `get_return_type_if_callable`.

That said, I'm very open to other suggestions.

Reply:
See above -- I like this, but my suggestion is to have it return `CallOutcome`, and keep all our logic for "simplifying" call outcomes consolidated in one place as methods on `CallOutcome`.